### PR TITLE
Nmt Model

### PR DESCRIPTION
**Pull request by okuchaiev**

### CODE REVIEW

File: nemo/collections/asr/models/configs/ctc_models_config.py
Comment (titu1994) [inline nemo/collections/asr/models/configs/ctc_models_config.py:81]: Since this is a model config, it doesnt need model: ..., all of the contents of `EncDecCTCConfig()` should be part of this since `EncDecCTCConfig()` is already a `ModelConfig`.

Two choices are - Rename this to `EncDecCTCModelPTConfig(model_cfg.NeMoModel)` or `EncDecCTCModelConfig(EncDecCTCConfig)`.

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py
Comment (titu1994) [inline nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py:53]: Add 
```python
        cfg = model_utils.convert_model_config_to_dict_config(cfg)
        cfg = model_utils.maybe_update_config_version(cfg)
```

Call the following utility methods from ModelPT before passing cfg. Assuming that cfg is a dataclass or a dictconfig is not guaranteed unless it passes through ModelPT constructor, so these two methods will force cfg to be valid

https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/models/rnnt_bpe_models.py#L65-L67

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py
Comment (titu1994) [inline nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py:110]: This is done by PTL properly just prior to training, not necessary to set explicitly in constructor

File: nemo/collections/nlp/modules/common/transformer/transformer_decoders.py
Comment (titu1994) [inline nemo/collections/nlp/modules/common/transformer/transformer_decoders.py:69]: Add before merge

File: nemo/collections/nlp/modules/common/transformer/transformer_generators.py
Comment (titu1994) [inline nemo/collections/nlp/modules/common/transformer/transformer_generators.py:166]: Add types before merge

File: nemo/core/classes/modelPT.py
Comment (titu1994) [inline nemo/core/classes/modelPT.py:33]: This should remain for easy importing by user.

File: nemo/core/config/modelPT.py
Comment (titu1994) [inline nemo/core/config/modelPT.py:66]: Add back HydraConfig

File: nemo/core/config/pytorch_lightning.py
Comment (titu1994) [inline nemo/core/config/pytorch_lightning.py:38]: Why is this necessary? Earlier we can pass the config to the constructor easily, now we need to use Instantiate and passing it in simply will raise an error in PTL trainer constructor.. 

This reduces visibility of arguments in constructor of Trainer (which may change while the config is not updated). Please revert.

File: nemo/utils/config_utils.py
Comment (titu1994) [inline nemo/utils/config_utils.py:23]: This is incorrect. The update method is built with the `NeMoModelConfig` template in mind, and will break if its passed the inner model. Please revert.

Comment (titu1994) [reply nemo/utils/config_utils.py:23]: Or if not, please update the code and helpers here to use the inner model rather inside the ModelConfig instead of assuming its a ModelPTConfig/NeMoModelConfig.

**Review by titu1994 — CHANGES_REQUESTED** [review]
Overall looks good. @ericharper There are some issues - specifically with `config_utils.py` and changing of the pytorch lightning trainer config to include `_target_`. Once those are resolved, lets merge.

Also, please remove NeMoModelConfig and rename the earlier ModelPTConfig to NeMoModelConfig instead.

File: Jenkinsfile
Comment (ericharper) [inline Jenkinsfile:780]: we should change this to 'main' ?

File: examples/cv/mnist_lenet5_image_classification_pure_lightning.py
Comment (ericharper) [inline examples/cv/mnist_lenet5_image_classification_pure_lightning.py:58]: We can revert this and just use `ptl.Trainer(**cfg.trainer)`

File: examples/nlp/machine_translation/conf/aayn_base.yaml
Comment (ericharper) [inline examples/nlp/machine_translation/conf/aayn_base.yaml:117]: this should be `hidden_size: ${model.decoder.hidden_size}`

File: nemo/utils/config_utils.py
Comment (okuchaiev) [reply nemo/utils/config_utils.py]: good catch, changed to NeMoConfig

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py
Comment (okuchaiev) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py:52]: added

File: nemo/collections/asr/models/configs/ctc_models_config.py
Comment (okuchaiev) [reply nemo/collections/asr/models/configs/ctc_models_config.py]: changed to class EncDecCTCModelConfig(model_cfg.NemoConfig):

File: examples/nlp/machine_translation/nmt_webapp/nmt_service.py
Comment (okuchaiev) [reply examples/nlp/machine_translation/nmt_webapp/nmt_service.py]: fixed

File: examples/nlp/machine_translation/conf/aayn_base.yaml
Comment (ericharper) [inline examples/nlp/machine_translation/conf/aayn_base.yaml:118]: This should be `decoder_vocab_size` but we don't have it anywhere in the config. I'm thinking that we should add 
```
decoder_embedding:
  vocab_size: 37000
```
And then interpolate this here:
```
  num_classes: ${model.decoder_embedding.vocab_size}
```

File: examples/nlp/machine_translation/enc_dec_nmt.py
Comment (ericharper) [inline examples/nlp/machine_translation/enc_dec_nmt.py:101]: We can revert this `instantiate`.

File: nemo/collections/nlp/data/machine_translation/machine_translation_dataset.py
Comment (ericharper) [inline nemo/collections/nlp/data/machine_translation/machine_translation_dataset.py:32]: Why `None` and not `MISSING`?

File: nemo/collections/nlp/modules/common/transformer/transformer_decoders.py
Comment (ericharper) [reply nemo/collections/nlp/modules/common/transformer/transformer_decoders.py:69]: I have another PR I'm working on where I'm adding this. Can we just add it there?

File: nemo/core/config/pytorch_lightning.py
Comment (ericharper) [reply nemo/core/config/pytorch_lightning.py:38]: We can revert it.

**Review by ericharper — APPROVED** [review]
LGTM. 

I think we can do a better job of interpolating in the configs, but we could also make those updates in a future PR.

This PR already is multiple PRs combined so I think it will be good if we get it merged and go back to smaller PRs more frequently on `main`.

File: nemo/collections/nlp/data/machine_translation/machine_translation_dataset.py
Comment (titu1994) [reply nemo/collections/nlp/data/machine_translation/machine_translation_dataset.py:32]: Yup this should be missing unless code can handle none case

File: nemo/collections/nlp/modules/common/transformer/transformer_decoders.py
Comment (titu1994) [reply nemo/collections/nlp/modules/common/transformer/transformer_decoders.py:69]: Sounds good

**Review by titu1994 — COMMENTED** [review]
Looks good, just ptl related revert needed, then let's merge. This PR is already massive so let's move the cleanup to subsequent prs.

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py
Comment (ericharper) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py:52]: @titu1994 do we need to call these at this level? can't we just call them at ModelPT level?

Comment (titu1994) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py:52]: ModelPT calls these two internally in it's constructor. However, this above line is being called before calling super() init method so ModelPT doesn't get the chance to convert cfg to a dictconfig

File: Jenkinsfile
Comment (titu1994) [reply Jenkinsfile:780]: +1

**Review by titu1994 — COMMENTED** [review]
Overall ready to approve and merge, lets address the last or Eric's comments and final review

**Comment by okuchaiev** [comment]
let's merge https://github.com/NVIDIA/NeMo/pull/1611 here first, then merge this one

File: nemo/collections/nlp/modules/common/transformer/transformer.py
Comment (titu1994) [inline nemo/collections/nlp/modules/common/transformer/transformer.py:53]: Do we need NM in the name?

File: nemo/collections/nlp/modules/common/transformer/transformer.py
Comment (titu1994) [inline nemo/collections/nlp/modules/common/transformer/transformer.py:109]: Do we need NM in the name?

File: nemo/collections/nlp/modules/common/transformer/transformer.py
Comment (titu1994) [inline nemo/collections/nlp/modules/common/transformer/transformer.py:154]: forward not typechecked

File: nemo/collections/nlp/modules/common/transformer/transformer.py
Comment (titu1994) [inline nemo/collections/nlp/modules/common/transformer/transformer.py:99]: forward not typechecked

File: nemo/collections/nlp/modules/common/transformer/transformer_encoders.py
Comment (titu1994) [inline nemo/collections/nlp/modules/common/transformer/transformer_encoders.py:86]: Avoid **kwargs everywhere throughout NeMo if possible

File: nemo/collections/nlp/modules/common/transformer/transformer_decoders.py
Comment (titu1994) [inline nemo/collections/nlp/modules/common/transformer/transformer_decoders.py:100]: Avoid **kwargs throughout NeMo as long as possible

File: nemo/core/classes/modelPT.py
Comment (titu1994) [reply nemo/core/classes/modelPT.py:33]: Please revert this line for easy import

File: nemo/core/config/modelPT.py
Comment (titu1994) [inline nemo/core/config/modelPT.py:66]: @ericharper Add hydra config back

File: nemo/core/config/pytorch_lightning.py
Comment (titu1994) [inline nemo/core/config/pytorch_lightning.py:38]: @ericharper  This should be reverted, signature must match PTL Trainer exactly. It is NemoConfig which should override this value to False..

**Review by titu1994 — COMMENTED** [review]
Minor semantic comments here and there, not blockers and can be cleaned up in subsequent PRs.

@ericharper There are two comments that must be fixed before merging this PR 
- Add back Hydra config to NemoConfig
- Revert logger: Any = False to logger: Any = True to remain exactly concistent with PTL. It is NemoModel which overrides logger value.

Final review to @ericharper for merge

**Review by titu1994 — APPROVED** [review]
LGTM** 
** @ericharper for final review and fix prior 2 requests

File: nemo/collections/nlp/modules/common/transformer/transformer.py
Comment (ericharper) [reply nemo/collections/nlp/modules/common/transformer/transformer.py:53]: `TransformerEncoder` is already taken by the `nn.module`. I don't know of a better suffix. Any ideas?

Comment (titu1994) [reply nemo/collections/nlp/modules/common/transformer/transformer.py:53]: Prepend NMT? If they are the same names, we should be merging functionality not duplicating classes, but that's out of scope for now. 

We don't have any NM modules anymore (they were there in <v0.11)

File: Jenkinsfile
Comment (ericharper) [reply Jenkinsfile:838]: I don't think so, exp_manager is set to null and the working directory is temporary, also we are not using cached datasets in this test.

File: nemo/collections/asr/models/configs/ctc_models_config.py
Comment (ericharper) [reply nemo/collections/asr/models/configs/ctc_models_config.py:80]: Done.

File: nemo/collections/nlp/data/machine_translation/machine_translation_dataset.py
Comment (ericharper) [reply nemo/collections/nlp/data/machine_translation/machine_translation_dataset.py:51]: Done.

File: nemo/collections/nlp/data/machine_translation/one_side_dataset.py
Comment (ericharper) [reply nemo/collections/nlp/data/machine_translation/one_side_dataset.py:30]: Done.

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_config.py
Comment (ericharper) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_config.py:47]: I put a todo there, let's fix in a future PR.

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py
Comment (ericharper) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py]: Done.

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py
Comment (ericharper) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py:149]: Removed it all together.

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py
Comment (ericharper) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py:171]: Done.

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py
Comment (ericharper) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py]: Done.

File: nemo/collections/nlp/modules/common/transformer/transformer.py
Comment (ericharper) [reply nemo/collections/nlp/modules/common/transformer/transformer.py:53]: The neural module TransformerEncoderNM takes ids as input. The ids then go through TransformerEmbedding and TransformerEncoder so that the out of TransformerEncoderNM is hidden states. So the functionality is not duplicated, this let's us swap encoders with HF and Megatron for example. We can come up with a better name soon I guess.

File: nemo/collections/nlp/modules/common/transformer/transformer.py
Comment (ericharper) [reply nemo/collections/nlp/modules/common/transformer/transformer.py:109]: See discussion above.

File: nemo/collections/nlp/modules/common/transformer/transformer.py
Comment (ericharper) [reply nemo/collections/nlp/modules/common/transformer/transformer.py:157]: Done.

File: nemo/collections/nlp/modules/common/transformer/transformer.py
Comment (ericharper) [reply nemo/collections/nlp/modules/common/transformer/transformer.py:101]: Done.

File: nemo/collections/nlp/modules/common/transformer/transformer_encoders.py
Comment (ericharper) [reply nemo/collections/nlp/modules/common/transformer/transformer_encoders.py:86]: Done.

File: nemo/core/config/modelPT.py
Comment (ericharper) [reply nemo/core/config/modelPT.py:66]: Done.

File: nemo/core/config/pytorch_lightning.py
Comment (ericharper) [reply nemo/core/config/pytorch_lightning.py:38]: Done.

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_config.py
Comment (titu1994) [reply nemo/collections/nlp/models/machine_translation/mt_enc_dec_config.py:48]: Sounds good. Leaving the comment unresolved to refer later what's left.

File: nemo/collections/nlp/modules/common/transformer/transformer.py
Comment (titu1994) [reply nemo/collections/nlp/modules/common/transformer/transformer.py:54]: Sure sounds good, future PR can rename it. Leaving comment open for later review.

**Comment by titu1994** [comment]
Thanks @ericharper for the fixes. LGTM, merge when ready and tests pass c

**Review by ericharper — APPROVED** [review]
Addressed change requests and set the upper bound of PTL to 1.1.3 due to accuracy regression.

**Pull request merged**
